ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
     }
   }
 
+  private def processPackageUpload(
+      entryId: DamlLogEntryId,
+      recordTime: Timestamp,
+      packageUploadEntry: DamlPackageUploadEntry,
+      inputState: Map[DamlStateKey, Option[DamlStateValue]]
+  ): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) = {
+    val submissionId = packageUploadEntry.getSubmissionId
+    def tracelog(msg: String): Unit =
+      logger.trace(
+        s"processPackageUpload[entryId=${prettyEntryId(entryId)}, submId=$submissionId]: $msg")
+
+    val archives = packageUploadEntry.getArchivesList.asScala
+    val packagesValidityResult =
+      archives.foldLeft[(Boolean, String)]((true, ""))(
+        (acc, archive) =>
+          if (archive.getPayload.isEmpty) (false, acc._2 ++ s"empty package '${archive.getHash}';")
```

Comment:
Is that enough to check the validity of the payload? Shouldn't you try to parse the archive?

Reply:
Well, the proper validation would go through a transitive closure of all the packages. It will have to be added at a later stage. We have agreed with Jussi to do the minimum at the moment. I can add a clear TODO in the code